### PR TITLE
Optimize diagnostics file management

### PR DESCRIPTION
**Pull request by tonidero**
### Description
This PR is the first part of a possible approach to avoid loading huge files in memory in diagnostics at once. Basically, in here we change to reading the file per lines instead of reading everything in one go.

#### Behavior changes
- Diagnostics will be Android 24+ only.
- An important change in behavior is that, instead of using the number of events in the diagnostics file to set a limit, we use the file size. So if on SDK configuration we detect that it's reached X size, we would clear the whole file and start over. 
- When we reach the file size limit, we remove the whole file, instead of just the oldest events.
- We remove the properties from the `MAX_EVENTS_STORED_LIMIT_REACHED` event since they don't make sense with the new approach. (This will require backend changes)
- Now we will have a limit to the max number of diagnostics events in memory and per request.

### CODE REVIEW

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt [hunk 1]
```diff
@@ -25,29 +32,36 @@ internal class FileHelper(
...
 
-    fun readFilePerLines(filePath: String): List<String> {
-        val readLines = mutableListOf<String>()
-        val file = getFileInFilesDir(filePath)
-        FileInputStream(file).use { fileInputStream ->
-            InputStreamReader(fileInputStream).use { inputStreamReader ->
-                BufferedReader(inputStreamReader).use { bufferedReader ->
-                    readLines.addAll(bufferedReader.readLines())
-                }
+    fun readFilePerLines(filePath: String, dataListener: DataListener<Pair<String, Int>>) {
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt:35]: I created this listener to be able to simulate a "Stream" of data with a completion.

BufferedReader provides the `lines` method: https://developer.android.com/reference/java/io/BufferedReader#lines() which would have been very useful, but it's API 24+ only... 

We could make diagnostics be API 24+ only, but I thought it would be good to gather data from older devices as well, so decided to just implement our own thing.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/diagnostics/DiagnosticsTracker.kt [hunk 2]
```diff
@@ -125,13 +125,10 @@ internal class DiagnosticsTracker(
         )
     }
 
-    fun trackMaxEventsStoredLimitReached(totalEventsStored: Int, eventsRemoved: Int, useCurrentThread: Boolean = true) {
+    fun trackMaxEventsStoredLimitReached(useCurrentThread: Boolean = true) {
         val event = DiagnosticsEntry.Event(
             name = DiagnosticsEventName.MAX_EVENTS_STORED_LIMIT_REACHED,
-            properties = mapOf(
-                "total_number_events_stored" to totalEventsStored,
-                "events_removed" to eventsRemoved,
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/diagnostics/DiagnosticsTracker.kt:133]: I thought these properties don't make sense anymore. We could send the file size, but in general, these files shouldn't get much bigger than the limit (currently of 500kb) so I thought it wouldn't be that useful.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt (hunk 1 above)
Comment (NachoSoto) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt:35]: Unless I was looking at it wrong, it seems like API 24+ has like 98% usage though?

Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt:35]: It would be 95.4%... It's pretty high, but not sure if we want to introduce that limitation on diagnostics... I think it's not that hard to support older clients so I would leave it as is. Lmk if you feel strongly about it though

image

Comment (NachoSoto) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt:35]: The thing is, that's 5% of _current_ users, possibly using old app versions. How many of those are still getting updates?
For comparison, I'm making diagnostics on iOS iOS 13.0+ only which came out in 2019. Level 24 is 2016, right?

Just saying that if it makes code easier, I doubt we'd even notice the difference by bumping the diagnostics API requirement.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt [hunk 3]
```diff
@@ -25,29 +34,26 @@ internal class FileHelper(
...
-        val file = getFileInFilesDir(filePath)
-        FileInputStream(file).use { fileInputStream ->
-            InputStreamReader(fileInputStream).use { inputStreamReader ->
-                BufferedReader(inputStreamReader).use { bufferedReader ->
-                    readLines.addAll(bufferedReader.readLines())
-                }
-            }
+    // This is using a lambda with a Stream instead of returning the Stream itself. This is so we keep
+    // the responsibility of closing the bufferedReader to this class. Note that the Stream should
+    // be used synchronously, otherwise the bufferedReader will be closed before the stream is used.
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt:39]: This is something that I don't fully like... But I didn't want to leak the responsibility of closing the readers.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/diagnostics/DiagnosticsSynchronizer.kt [hunk 4]
```diff
@@ -91,15 +103,11 @@ internal class DiagnosticsSynchronizer(
...
-        val diagnosticsList = diagnosticsFileHelper.readDiagnosticsFile()
-        val diagnosticsInFileCount = diagnosticsList.size
-        if (diagnosticsInFileCount > MAX_NUMBER_EVENTS) {
-            val eventsToRemoveCount = diagnosticsInFileCount - MAX_NUMBER_EVENTS + 1
-            diagnosticsFileHelper.deleteOlderDiagnostics(eventsToRemoveCount)
-            diagnosticsTracker.trackMaxEventsStoredLimitReached(diagnosticsInFileCount, eventsToRemoveCount)
-            return diagnosticsFileHelper.readDiagnosticsFile()
+        var eventsToSync: List<JSONObject> = emptyList()
+        diagnosticsFileHelper.readDiagnosticsFile { stream ->
+            eventsToSync = stream.limit(MAX_EVENTS_TO_SYNC_PER_REQUEST).collect(Collectors.toList())
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/diagnostics/DiagnosticsSynchronizer.kt:108]: Adding a limit to the events in memory/per request was done previously in a separate PR, but I decided to move it to this PR for simplicity

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt (hunk 3 above)
Comment (NachoSoto) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/FileHelper.kt:39]: It makes sense tho

File: purchases/src/test/java/com/revenuecat/purchases/common/diagnostics/DiagnosticsSynchronizerTest.kt [hunk 5]
```diff
@@ -58,15 +59,49 @@ class DiagnosticsSynchronizerTest {
...
+        every { diagnosticsFileHelper.isDiagnosticsFileTooBig() } returns true
+        every { diagnosticsTracker.trackMaxEventsStoredLimitReached() } just Runs
+
+        diagnosticsSynchronizer.clearDiagnosticsFileIfTooBig()
+
+        verify(exactly = 1) { diagnosticsTracker.trackMaxEventsStoredLimitReached() }
+    }
+
+    @Test
+    fun `clearDiagnosticsFileIfTooBig clears diagnostics file if dianostics file too big`() {
```
Comment (NachoSoto) [inline purchases/src/test/java/com/revenuecat/purchases/common/diagnostics/DiagnosticsSynchronizerTest.kt:85]: What do you think about a more "functional" test that adds say 500 events and verifies that it's not considered too big?
As a way to validate that the size limit allows for some number of events that we're happy with.

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesConfiguration.kt [hunk 6]
```diff
@@ -74,6 +74,8 @@ open class PurchasesConfiguration(builder: Builder) {
          * Examples of this information include response times, cache hits or error codes.
          * This information will be anonymized so it can't be traced back to the end-user.
          * The default value is false.
+         *
+         * Diagnostics is only available in Android API 24+
```
Comment (NachoSoto) [inline purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesConfiguration.kt:78]: Do we need to put this in the docs too?

**Review by NachoSoto — APPROVED** [review]
Sorry for the delay!

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesConfiguration.kt (hunk 6 above)
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesConfiguration.kt:78]: We haven't added diagnostics to the docs yet since we were planning to add extra functionality before that, so nothing to update

File: purchases/src/test/java/com/revenuecat/purchases/common/diagnostics/DiagnosticsSynchronizerTest.kt (hunk 5 above)
Comment (tonidero) [reply purchases/src/test/java/com/revenuecat/purchases/common/diagnostics/DiagnosticsSynchronizerTest.kt:85]: Good idea, done in https://github.com/RevenueCat/purchases-android/pull/1194/commits/265d1f66dc9f1b65d96f569b1c379d8d6ed88921
